src/iminuit/cost.py

```diff
@@ -107,6 +79,141 @@ def _soft_l1_cost(y, ye, ym):
     return _soft_l1_loss(_z_squared(y, ye, ym))
 
 
+def chi2(y, ye, ym):
+    """
+    Compute (potentially) chi2-distributed cost.
+
+    The value returned by this function is chi2-distributed, if the observed values are
+    normally distributed around the expected values with the provided standard deviations.
+
+    Parameters
+    ----------
+    y : array-like
+        Observed values.
+    ye : array-like
+        Uncertainties of values.
+    ym : array-like
+        Expected values.
+
+    Returns
+    -------
+    float
+        Const function value.
+    """
+    y, ye, ym = np.atleast_1d(y, ye, ym)
+    return np.sum(_z_squared(y, ye, ym))
+
+
+def multinominal_chi2(n, p):
+    """
+    Compute asymptotically chi2-distributed cost for binomially-distributed data.
+
+    Parameters
+    ----------
+    n : array-like
+        Observed counts.
+    p : array-like
+        Probabilities for an event to be end up in each bin.
+        Must satisfy sum(p) == 1.
+
+    Returns
+    -------
+    float
+        Cost function value.
+
+    Notes
+    -----
+    The implementation makes the result asymptotically chi2-distributed and
+    keeps the sum small near the minimum, which helps to maximise the numerical
+    accuracy for Minuit.
+
+    The formula is derived from the likelihood ratio with respect to a
+    saturated model (which has mu = n):
+
+    Q = -2 (lnP - lnP')
+    lnP = ln n! - sum_i k_i! + sum_i k_i ln(p_i)
+    lnP' = ln n! - sum_i k_i! + sum_i k_i ln(k_i / n)
+    Q = -2 sum_i k_i (ln(p_i) - ln(k_i / n))
+      = -2 sum_i k_i (ln(mu_i) - ln(n) - ln(k_i) + ln(n))
+      = 2 sum_i k_i (ln(k_i) - ln(mu_i))
+    """
+    n = np.atleast_1d(n)
+    p = np.atleast_1d(p)
+    mu = np.sum(n) * p
+    return 2 * np.sum(n * (_safe_log(n) - _safe_log(mu)))
+
+
+def poisson_chi2(n, mu):
+    """
+    Compute asymptotically chi2-distributed cost for Poisson-distributed data.
+
+    Parameters
+    ----------
+    n : array-like
+        Observed counts.
+    mu : array-like
+        Expected counts.
+
+    Returns
+    -------
+    float
+        Cost function value.
+
+    Notes
+    -----
+    The implementation makes the result asymptotically chi2-distributed and
+    keeps the sum small near the minimum, which helps to maximise the numerical
+    accuracy for Minuit.
+
+    The formula is derived from the likelihood ratio with respect to a
+    saturated model (which has mu = n):
+
+    Q = -2 (lnP - lnP')
+    lnP = sum_i -mu_i + k_i ln(mu_i) - ln(k_i!)
+    lnP' = sum_i -k_i + k_i ln(k_i) - ln(k_i!)
+    Q = 2 sum_i mu_i - k_i + k_i (ln(k_i) - ln(mu_i))
+    """
+    n = np.atleast_1d(n)
+    mu = np.atleast_1d(mu)
+    return 2 * np.sum(mu - n + n * (_safe_log(n) - _safe_log(mu)))
+
+
+def barlow_beeston_lite_chi2(n, mu, mu_var):
```

Comment:
@AhmedAbdelmotteleb I slightly generalised and documented your function, could you please check whether this is correct?